### PR TITLE
Network Observability FlowCollector API doc updates

### PR DESCRIPTION
**Pull request by skrthomas**
Version(s):

4.11+
Issue:

https://issues.redhat.com/browse/OSDOCS-7660

Link to docs preview:

https://64761--docspreview.netlify.app/openshift-enterprise/latest/networking/network_observability/flowcollector-api
QE review:
- [X] QE has approved this change. 

Additional information:

Doc generated by https://github.com/netobserv/network-observability-operator/pull/412

### CODE REVIEW

**Comment by skrthomas** [comment]
@jotak I think the preview looks good. Good to go in my view! 
@memodi can you take a look at this for QE approval? I think all these parameters have already been ACK'd in other engineering PRs.

**Review by jotak — APPROVED** [review]
LGTM!

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 1]
```diff
@@ -59,17 +57,14 @@ Type::
 Description::
 +
 --
-`FlowCollectorSpec` defines the desired state of the FlowCollector resource.  +
+Defines the desired state of the FlowCollector resource.  +
  +
- *: the mention of _"unsupported"_, or _"deprecated"_ for a feature throughout this document means that this feature is not officially supported by Red Hat. It might have been, for instance, contributed by the community and accepted without a formal agreement for maintenance. The product maintainers might provide some support for these features as a best effort only.
+ *: the mention of "unsupported", or "deprecated" for a feature throughout this document means that this feature is not officially supported by Red Hat. It might have been, for instance, contributed by the community and accepted without a formal agreement for maintenance. The product maintainers might provide some support for these features as a best effort only.
```
Comment (memodi) [inline modules/network-observability-flowcollector-api-specifications.adoc:62]: any reason why we mention this multiple times?

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 2]
```diff
@@ -551,22 +556,22 @@ Required::
...
+| IPFIX configuration, such as the IP address and port to send enriched IPFIX flows to. [Unsupported (*)].
 
 | `kafka`
 | `object`
 | Kafka configuration, such as the address and topic, to send enriched flows to.
 
 | `type`
 | `string`
-| `type` selects the type of exporters. The available options are `KAFKA` and `IPFIX`. `IPFIX` is _unsupported (*)_.
+| `type` selects the type of exporters. The available options are `KAFKA` and `IPFIX`. `IPFIX` is unsupported (*).
```
Comment (memodi) [inline modules/network-observability-flowcollector-api-specifications.adoc:567]: should it be specified as `[Unsupported (*)].` it's inconsistent.

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 3]
```diff
@@ -1335,7 +1552,7 @@ Type::
 
 | `ignoreTags`
 | `array (string)`
-| `ignoreTags` is a list of tags to specify which metrics to ignore. Each metric is associated with a list of tags. More details in https://github.com/netobserv/network-observability-operator/tree/main/controllers/flowlogspipeline/metrics_definitions . Available tags are: `egress`, `ingress`, `flows`, `bytes`, `packets`, `namespaces`, `nodes`, `workloads`.
+| `ignoreTags` is a list of tags to specify which metrics to ignore. Each metric is associated with a list of tags. More details in https://github.com/netobserv/network-observability-operator/tree/main/controllers/flowlogspipeline/metrics_definitions . Available tags are: `egress`, `ingress`, `flows`, `bytes`, `packets`, `namespaces`, `nodes`, `workloads`, `nodes-flows`, `namespaces-flows`, `workloads-flows`. Namespace-based metrics are covered by both `workloads` and `namespaces` tags, hence it is recommended to always ignore one of them (`workloads` offering a finer granularity).
```
Comment (memodi) [inline modules/network-observability-flowcollector-api-specifications.adoc:1555]: did we add *flows metrics in 1.4? I thought we deferred that feature to 1.5? cc @jpinsonneau

**Review by memodi — CHANGES_REQUESTED** [review]
overall LGTM just few minor comments/questions

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 2 above)
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:567]: @jotak is IPFIX supported for 1.4 release time or is it going out in 4.14 time?

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 1 above)
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:62]: This is basically our boilerplate blurb whenever there are unsupported/deprecated items that are listed in the API documentation, so it appears at each one of those. Just because, if we put an asterisk and then refer to a single blurb on the page, the user has to go all the way to the bottom of the page to find the blurb then go back to where they were. So we just mention it at each point that its relevant.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 2 above)
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:567]: Via slack, we discussed IPFIX is supported so this part of the API documentation will need to be updated cc @jotak

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 3 above)
Comment (jpinsonneau) [reply modules/network-observability-flowcollector-api-specifications.adoc:1555]: It is part of 1.4 yes:
```
// `ignoreTags` is a list of tags to specify which metrics to ignore. Each metric is associated with a list of tags. More details in https://github.com/netobserv/network-observability-operator/tree/main/controllers/flowlogspipeline/metrics_definitions .
	// Available tags are: `egress`, `ingress`, `flows`, `bytes`, `packets`, `namespaces`, `nodes`, `workloads`, `nodes-flows`, `namespaces-flows`, `workloads-flows`.
	// Namespace-based metrics are covered by both `workloads` and `namespaces` tags, hence it is recommended to always ignore one of them (`workloads` offering a finer granularity).
	//+kubebuilder:default:={"egress","packets","nodes-flows","namespaces-flows","workloads-flows","namespaces"}
	// +optional
	IgnoreTags []string `json:"ignoreTags"`
```

https://github.com/netobserv/network-observability-operator/blob/release-1.4/api/v1beta1/flowcollector_types.go#L350-L355

Adding `flows` to your list will ignore node / namespace / workload flows health graphs. It's just a shortcut.

Comment (jpinsonneau) [reply modules/network-observability-flowcollector-api-specifications.adoc:1555]: To clarify, the postponed part is only [NETOBSERV-1017](https://issues.redhat.com/browse/NETOBSERV-1017) about adding `resources` tag for pods & services metrics.
This will require [NETOBSERV-1284](https://issues.redhat.com/browse/NETOBSERV-1284) `white-list capability` first to avoid any mistakes on client side while manipulating these tags.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 2 above)
Comment (jotak) [reply modules/network-observability-flowcollector-api-specifications.adoc:567]: oops, good catch, I'll update the upstream doc

Comment (jotak) [reply modules/network-observability-flowcollector-api-specifications.adoc:567]: => https://github.com/netobserv/network-observability-operator/pull/426

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 3 above)
Comment (memodi) [reply modules/network-observability-flowcollector-api-specifications.adoc:1555]: oh, I completely missed those new tags while testing 1075 changes and while testing it now found a bug: https://issues.redhat.com/browse/NETOBSERV-1315

Comment (jpinsonneau) [reply modules/network-observability-flowcollector-api-specifications.adoc:1555]: Glad we had this thread then :laughing: I'm checking the bug right now. Thanks @memodi !

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 4]
```diff
@@ -619,20 +624,122 @@ Required::
...
+| `string`
+| Name of the config map or secret containing the file
+
+| `namespace`
+| `string`
+| Namespace of the config map or secret containing the file. If omitted, the default is to use the same namespace as where Network Observability is deployed. If the namespace is different, the config map or the secret is copied so that it can be mounted as required.
+
+| `type`
+| `string`
+| Type for the file reference: "configmap" or "secret"
```
Comment (stevsmit) [inline modules/network-observability-flowcollector-api-specifications.adoc:701]: Would these strings be in backticks since they're values?

**Review by stevsmit — COMMENTED** [review]
@skrthomas I think overall this looks really good. API docs are inherently weird, in that the normal conventions of OCP guidelines, imo, kind of go out the window; I just think it's difficult to provide a thorough review using those guidelines. 

That said, broadly, I would take a look at the rendering on the portal, and specifically at *Descriptions*, and consider how some entries, like Sampling, are not in backticks in the description (also, you have two entries for *Sampling*, and have them formatted in two different ways. I'd choose the sampling field under ".spec.agent.ipfix" personally). I see why you did it like this, and I think it's fine, but I also could see it working as something like what follows:

`sampling` sets the rate of the flow reporter. 

Namespace is another field that doens't follow the same pattern. Just fyi. Enable is another one. That said, I think it's fine as-is, and hard to follow the same conventions for some field; I'm just pointing this out. 

Another thing to consider might be putting literal values in backticks, as per the OCP Guidelines, for example, `9093`. I believe that this falls under the category of literal value. Would things like GOGC and GOMAXPROCS also be values for the `env` field? Any "value" that can be set under a parameter, that is mention in the Descriptions field, probably needs set in backticks (e.g., port has a few)

Overall, good work. These are a log to get through.

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 5]
```diff
@@ -25,21 +23,21 @@ Type::
 
 | `apiVersion`
 | `string`
-| APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and might reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+| APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
```
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:26]: This is a direct copy of asciidoc generated from the source code...I'm not sure why these got changed.

[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:26]: Maybe they were holdovers I noticed last round of API updates and I manually changed them but didn't tell Joel to change them on his end.

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 6]
```diff
@@ -25,21 +23,21 @@ Type::
...
 | `object`
 | Standard object's metadata. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
 
 | `spec`
 | `object`
-| `FlowCollectorSpec` defines the desired state of the FlowCollector resource.  +
+| Defines the desired state of the FlowCollector resource.  +
  +
- *: the mention of _"unsupported"_, or _"deprecated"_ for a feature throughout this document means that this feature is not officially supported by Red Hat. It might have been, for instance, contributed by the community and accepted without a formal agreement for maintenance. The product maintainers might provide some support for these features as a best effort only.
+ *: the mention of "unsupported", or "deprecated" for a feature throughout this document means that this feature is not officially supported by Red Hat. It might have been, for instance, contributed by the community and accepted without a formal agreement for maintenance. The product maintainers might provide some support for these features as a best effort only.
```
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:40]: This full statement only appears in the `spec` table and the `spec` field description...The asterisks is used throughout but the text itself is only at the beginning. When I commented to Mehul earlier, I was thinking that this whole statement was repeated, but its only "deprecated" or "unsupported" that appear beside the "*". I think its honestly fine like this

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 7]
```diff
@@ -179,27 +172,35 @@ Type::
 
 | `excludeInterfaces`
 | `array (string)`
-| `excludeInterfaces` contains the interface names that will be excluded from flow tracing. An entry is enclosed by slashes, such as `/br-/`, is matched as a regular expression. Otherwise it is matched as a case-sensitive string.
+| `excludeInterfaces` contains the interface names that are excluded from flow tracing. An entry is enclosed by slashes, such as `/br-/`, is matched as a regular expression. Otherwise it is matched as a case-sensitive string.
```
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:175]: I also don't think the comma is needed before this "and". It should be "...`/br-/` and is matched as a reg..."

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 8]
```diff
@@ -914,6 +1123,10 @@ Type::
 | `string`
 | `batchWait` is the maximum time to wait before sending a batch.
 
+| `enable`
+| `boolean`
+| enable storing flows to Loki. It is required for the {product-title} Console plugin installation.
```
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:1128]: Good suggestion.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:40]: On the currently live customer portal, I see the full text with asterisks and that looks okay to me: 
image

[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:40]: The portal chapters do kinda make this look weird.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:62]: I responded to the earlier comment with a screencapture from the live customer portal, where I see this entire statement with the asterisks and it looks OK to me. I can swap "for instance" for "for example" in both cases of this deprecation stmt.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 4 above)
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:701]: Strings are in quotes I think, https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#yaml-formatting-for-kubernetes-and-openshift-api-objects

IN the YAML example, I see "" around string values. It isn't actually clearly defined there.

Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:701]: Unless I'm misinterpreting this.

**Comment by skrthomas** [comment]
Thanks for the review @stevsmit !! 

- Sampling - I agree, in `spec.agent.ebpf.samling`, in that description, I think we can say "`sampling` sets the rate of the flow reporter" cc @jotak 
- I'm unsure about putting GOGC and GOMAXPROCS in backticks...Its only used in the NetObserv API reference, so I don't see any other examples of the treatment in other areas of docs.openshift. I put `9093` in backticks though. 
- I changed `enable` as you suggested in your comment. There are a lot of namespaces...yeah. Its the main inconsistent one. At least its consistently inconsistent? 😅  I'm ok leaving it...we could say something like "`namespace` designates where the config map, etc"

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 9]
```diff
@@ -25,21 +23,21 @@ Type::
 
 | `apiVersion`
 | `string`
-| APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and might reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
+| APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources
 
 | `kind`
 | `string`
-| Kind is a string value representing the REST resource this object represents. Servers might infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
```
[the remark this answers is not in the document]
Comment (jotak) [reply modules/network-observability-flowcollector-api-specifications.adoc]: that's also upstream k8s

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 10]
```diff
@@ -25,21 +23,21 @@ Type::
...
+| Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
```
[the remark this answers is not in the document]
Comment (jotak) [reply modules/network-observability-flowcollector-api-specifications.adoc]: that's also upstream k8s

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (jotak) [reply modules/network-observability-flowcollector-api-specifications.adoc]: In fact, the first "is" shouldn't be there and I guess that's misleading you for getting the meaning here. It should be like:

> An entry enclosed by slashes, such as `/br-/`, is matched as a regular expression.

Side-question: any preference between "enclosed by", "enclosed in" or "enclosed within" ?

File: modules/network-observability-flowcollector-api-specifications.adoc [hunk 11]
```diff
@@ -514,7 +519,7 @@ Type::
 
 | `requests`
 | `integer-or-string`
-| Requests describes the minimum amount of compute resources required. If Requests is omitted for a container, it defaults to Limits if that is explicitly specified, otherwise to an implementation-defined value. More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+| Requests describes the minimum amount of compute resources required. If Requests is omitted for a container, it defaults to Limits if that is explicitly specified, otherwise to an implementation-defined value. Requests cannot exceed Limits. More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
```
[the remark this answers is not in the document]
Comment (jotak) [reply modules/network-observability-flowcollector-api-specifications.adoc:522]: this part is taken from k8s upstream

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc]: @jotak Can we say 
> Set `enable` to `true` to store flows in Loki.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc:522]: Ok, thanks for confirming. I did not add any backticks to this on my side.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (skrthomas) [reply modules/network-observability-flowcollector-api-specifications.adoc]: @jotak I think "enclosed by" is good.

File: modules/network-observability-flowcollector-api-specifications.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (jotak) [reply modules/network-observability-flowcollector-api-specifications.adoc]: @skrthomas sounds good to me, i'm also updating upstream
